lib/DiscoveryStreamFeed.jsm

```diff
@@ -163,25 +163,51 @@ this.DiscoveryStreamFeed = class DiscoveryStreamFeed {
     }
   }
 
+  buildFeedPromises(layout, isStartup) {
+    const initialData = {
+      newFeedsPromises: [],
+      newFeeds: {},
+    };
+
+    return layout.filter(row => row && row.components).reduce(({newFeedsPromises, newFeeds}, row) => {
+      row.components.filter(component => component && component.feed).forEach(component => {
+        const {url} = component.feed;
```

Comment:
I find buildFeedPromises very difficult to read and reason about, and I think in large part it's because five layers of abstraction are introduced in the above three lines (return, filter, reduce, filter, forEach).  This might want to broken up into multiple real functions, but at the very least, I'd suggest not doing nearly so much with manually in-lined arrow functions.

Reply:
https://docs.codeclimate.com/docs/cognitive-complexity has a some discussion around this.